parser/types/field_type.go

```diff
@@ -121,10 +127,43 @@ func (ft *FieldType) SetFlen(flen int) {
 	ft.flen = flen
 }
 
+func (ft *FieldType) SetFlenUnderLimit(flen int) {
+	if ft.tp == mysql.TypeNewDecimal {
+		ft.flen = mathutil.Min(flen, mysql.MaxDecimalWidth)
```

Comment:
need to check if  flen > 0 ?

Reply:
it would be -1 if unknown.

Reply:
But [here](https://github.com/pingcap/tidb/pull/34399/files#diff-22b102f6cca43864b03a04549066f46bf67e810e89f87271cfdad1e1c6c17bb9R190) set the flen to `MaxDecimalWidth` if flen < 0?

Reply:
that is an agg function, where the flen and decimal should be valid, instead of -1.

Reply:
in other words, this PR does not get rid of -1 cases, but be under the max values.